### DISCUSSION TITLE
POSIX compatible command line argument handling

### DISCUSSION THREAD

**Discussion by FrankRay78** [spectreconsole/spectre.console#1025] [General]
There are several bugs raised covering various edge cases and/or platform specific differences with how command line arguments are parsed and handled by spectre.console. 

Discussions about POSIX compatible command line argument handling can be done here (in addition to issue specific commentary).

- [Invalid handling of parameter values starting with double quotes ("), e.g. --order-by='"quoted" value'](https://github.com/spectreconsole/spectre.console/issues/891)
- [POSIX: Invalid handling of parameter values starting with double quotes ("), e.g. --order-by='"quoted" value'](https://github.com/spectreconsole/spectre.console/issues/959)
- [Value starting with - is mistreated as extra option](https://github.com/spectreconsole/spectre.console/issues/890)
- [Wrong handling of empty string and space values (e.g. --order-by="" or --order-by=" ")](https://github.com/spectreconsole/spectre.console/issues/842)
- [A boolean flag followed by a positional argument leads the the error Flags cannot be assigned a value.](https://github.com/spectreconsole/spectre.console/issues/825)
- [Support options before arguments](https://github.com/spectreconsole/spectre.console/issues/193)

**Comment by FrankRay78** [comment]
> Would it help, if I compose a set of edge case parameter sets (e.g. --option '"value" test') compile the following code as x64 for Windows, macOS and Linux and provide the output of these edge cases on all platforms?:

Feel free to summarise your findings here @sandreas.

**Comment by sandreas** [comment]
Ok, let's regard the edge cases below for a first basic test, assuming that Windows and Linux/macOS (POSIX)  handle arguments a bit different:

- Windows does not allow `'` as enclosure in `cmd` (I don't know, how Powershell will handle this), so this will produce 3 arguments, instead of one (`--order-by 'spaced value'` => `args[0]=--order-by`,   `args[1]='spaced`, `args[2]=value'`) while `--order-by "spaced value"` will prodce only 2: `--order-by` and `spaced value`)
- POSIX does allow either `'`or `"` as enclosure, while `""` allows variables to be expanded (e.g. `"$HOME/downloads"`) while `'` does not
- POSIX additionaly expands chars like `*` before handing it over to the command line tool, so `--select *` would produce a different result from `--select '*'`
- POSIX also reduces `"` under specific circumstances (e.g. `echo --order-by="size"` results in `--order-by=size`)

```bash
# OPTIONS
--order-by "-size"
--order-by ""
--order-by " "
## short form
-o-size # => --order-by "-size"
-o "" # => --order-by ""
-o"" # => --order-by ""
-o" " # => --order-by " "
## POSIX only because windows does not support single quote enclosures
--order-by '"something quoted"'
--order-by '"something quoted" with additional text'

# FLAGS (+ARGUMENTS)
--dry-run ./directory
-df # --debug, --force

# FLAG / OPTION combinations (is that a wanted behaviour?)
-dfo "-size" # --debug --force --order-by "-size"
```

If someone can think of more useful edge cases, feel free to add a comment...
